src/types.rs

```diff
@@ -330,3 +330,37 @@ impl DestinationSlot {
         self.dest.get()
     }
 }
+
+/// multishot header struct
+/// multishot recvmsg layout: <header><name><control><payload>
+/// https://lwn.net/Articles/901221/ 
+#[derive(Default, Debug, Clone, Copy)]
+#[repr(transparent)]
+pub struct RecvmsgOut(sys::io_uring_recvmsg_out);
```

Comment:
I was playing with multishot recvmsg (see https://github.com/tokio-rs/io-uring/issues/187#issuecomment-1426993071) and from a consumer point-of-view I have the feeling this is not the most useful data-structure to have.

What I'm currently playing with instead looks like this:
```rust
pub struct RecvMsgOut<'buf> {
    buf: &'buf [u8],
    header: sys::io_uring_recvmsg_out,
    msghdr_name_len: u32,
    msghdr_control_len: u32,
}

impl<'buf> RecvMsgOut<'buf> {
    pub fn parse(buf: &'buf [u8], name_len: u32, control_len: u32) -> Result<Self> { ... }
}
```

In plain words, my current understanding is that to properly handle the result of multishot recvmsg the consumer needs to glue and store together all of the following:
 * the raw data output in the provided buffer
 * the `io_uring_recvmsg_out` content in that data (validated and parsed upfront)
 * the `namelen` and `controllen` that were provided in the SQE `msghdr` (not the whole struct content, just those two are strictly required)

If you are not in a hurry, I can try to finish writing tests and submit what I have next week.

Reply:
Thank you both. As I just wrote in #187, I look forward to the work in @lucab 's branch, sounds like a lot was figured out already and we will all learn a lot looking at the unit test.

@glenn-wang Thank you for still being interested. I don't know if you know this but this `io-uring` branch is not generally meant for writing a final application but that doesn't mean it can't be done. In fact, I can imagine there are some proprietary solutions already that use this crate with either std threads or their own choice of async runtime.

You can consider putting this in draft mode, we aren't likely to be using the data structure.

I don't know where your past experience lies, but if you have any experience with qemu test runners, I know the maintainer of this repo has asked for help in getting the CI run with a newer version of the kernel. A 6.1 version would be very good, with a way to get it rebuilt as it is now considered an LTS with quite a few years of support planned - so any bug fixes to the io_uring device code will be reflected in rebuilds. I don't think new features will be back ported to it though, so that's a bummer but understandable.

Reply:
> I don't know if you know this but this io-uring branch is not generally meant for writing a final application but that doesn't mean it can't be done. In fact, I can imagine there are some proprietary solutions already that use this crate with either std threads or their own choice of async runtime.

I know this. For example , [tokio-rs/tokio-uring](https://github.com/tokio-rs/tokio-uring) is very good. But that project may not encapsulate all the features of io-uring, and it is lagging behind of tokio-rs/io-uring.

> I don't know where your past experience lies, but if you have any experience with qemu test runners, I know the maintainer of this repo has asked for help in getting the CI run with a newer version of the kernel. 

I have no experience of qemu test runners so have no ability to make a help.

Reply:
Okay. Didn't mean to offend. You are welcome to use this repo for your own purposes of course.

Reply:
The work-in-progress logic I was talking about is now up at https://github.com/tokio-rs/io-uring/pull/192.